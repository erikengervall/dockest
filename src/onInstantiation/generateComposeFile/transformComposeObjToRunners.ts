import { RedisRunner, PostgresRunner, GeneralPurposeRunner, ZooKeeperRunner, KafkaRunner } from '../../runners'
import { DockestConfig } from '../../index'
import { ComposeFile } from '../../runners/@types'
import ConfigurationError from '../../errors/ConfigurationError'

const { keys } = Object

const REDIS_REG_EXP = RegExp('redis')
const POSTGRES_REG_EXP = RegExp('postgres')
const ZOOKEEPER_REG_EXP = RegExp('zookeeper')
const KAFKA_REG_EXP = RegExp('kafka')

export default (config: DockestConfig, composeObj: ComposeFile) =>
  keys(composeObj.services).map(service => {
    const attachedRunnerExist = config.runners.find(runner => runner.runnerConfig.service === service)
    if (attachedRunnerExist) {
      return attachedRunnerExist
    }

    const composeService = composeObj.services[service]

    /**
     * Redis
     */
    if (REDIS_REG_EXP.test(composeService.image || '') || REDIS_REG_EXP.test(service)) {
      return new RedisRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
      })
    }

    /**
     * Postgres
     */
    if (POSTGRES_REG_EXP.test(composeService.image || '') || POSTGRES_REG_EXP.test(service)) {
      if (!composeService.environment) {
        throw new ConfigurationError(`${service}: Invalid environment`)
      }

      return new PostgresRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
        database: `${composeService.environment[PostgresRunner.ENVIRONMENT_DATABASE]}`,
        password: `${composeService.environment[PostgresRunner.ENVIRONMENT_PASSWORD]}`,
        username: `${composeService.environment[PostgresRunner.ENVIRONMENT_USERNAME]}`,
      })
    }

    /**
     * ZooKeeper
     */
    if (ZOOKEEPER_REG_EXP.test(composeService.image || '') || ZOOKEEPER_REG_EXP.test(service)) {
      if (!composeService.environment) {
        throw new ConfigurationError(`${service}: Invalid environment`)
      }

      return new ZooKeeperRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
      })
    }

    /**
     * Kafka
     */
    if (KAFKA_REG_EXP.test(composeService.image || '') || KAFKA_REG_EXP.test(service)) {
      if (!composeService.environment) {
        throw new ConfigurationError(`${service}: Invalid environment`)
      }

      return new KafkaRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
      })
    }

    return new GeneralPurposeRunner({
      service,
      image: composeService.image,
    })
  })
