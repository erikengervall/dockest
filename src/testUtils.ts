import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
