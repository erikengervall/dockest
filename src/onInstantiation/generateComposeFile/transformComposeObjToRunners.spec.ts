import transformComposeObjToRunners from './transformComposeObjToRunners'
import testUtils from '../../testUtils'
import { singlePostgres, postgresAndRedis } from '../../../fixtures/composeObj'

const { createDockestConfig, initializedRunners } = testUtils({ realLoggers: true })

describe('transformComposeObjToRunners', () => {
  it('should detect conflict and use the already existing runner', () => {
    const dockestConfig = createDockestConfig({
      runners: Object.values(initializedRunners),
      realLoggers: true,
    })

    const result = transformComposeObjToRunners(dockestConfig, singlePostgres)

    expect(result).toMatchSnapshot()
  })

  it('should detect conflict and use the already existing runner', () => {
    const dockestConfig = createDockestConfig({
      runners: [initializedRunners.redisRunner],
      realLoggers: true,
    })

    const result = transformComposeObjToRunners(dockestConfig, postgresAndRedis)

    expect(result).toMatchSnapshot()
  })
})
