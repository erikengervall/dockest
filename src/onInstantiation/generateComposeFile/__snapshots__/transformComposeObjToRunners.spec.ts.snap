// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformComposeObjToRunners should detect conflict and use the already existing runner 1`] = `
Array [
  PostgresRunner {
    "containerId": "",
    "createResponsivenessCheckCmd": [Function],
    "getComposeService": [Function],
    "initializer": "",
    "logger": Logger {
      "debug": [Function],
      "error": [Function],
      "info": [Function],
      "runnerService": "postgres",
      "runnerSymbol": "ðŸ¦‡ ",
      "setRunnerSymbol": [Function],
      "warn": [Function],
    },
    "runnerConfig": Object {
      "build": "",
      "commands": Array [],
      "connectionTimeout": 5,
      "database": "_",
      "dependsOn": Array [],
      "host": "localhost",
      "image": "postgres/image:123",
      "networks": undefined,
      "password": "_",
      "ports": Object {
        "5432": "5432",
      },
      "props": Object {},
      "responsivenessTimeout": 30,
      "service": "postgres",
      "username": "_",
    },
  },
]
`;

exports[`transformComposeObjToRunners should detect conflict and use the already existing runner 2`] = `
Array [
  PostgresRunner {
    "containerId": "",
    "createResponsivenessCheckCmd": [Function],
    "getComposeService": [Function],
    "initializer": "",
    "logger": Logger {
      "debug": [Function],
      "error": [Function],
      "info": [Function],
      "runnerService": "postgres",
      "runnerSymbol": "ðŸ¦‡ ",
      "setRunnerSymbol": [Function],
      "warn": [Function],
    },
    "runnerConfig": Object {
      "build": "",
      "commands": Array [],
      "connectionTimeout": 5,
      "database": "_",
      "dependsOn": Array [],
      "host": "localhost",
      "image": "postgres/image:123",
      "networks": undefined,
      "password": "_",
      "ports": Object {
        "5432": "5432",
      },
      "props": Object {},
      "responsivenessTimeout": 30,
      "service": "postgres",
      "username": "_",
    },
  },
  RedisRunner {
    "containerId": "",
    "createResponsivenessCheckCmd": [Function],
    "getComposeService": [Function],
    "initializer": "",
    "logger": Logger {
      "debug": [Function],
      "error": [Function],
      "info": [Function],
      "runnerService": "redis",
      "runnerSymbol": "ðŸ¦‡ ",
      "setRunnerSymbol": [Function],
      "warn": [Function],
    },
    "runnerConfig": Object {
      "build": "",
      "commands": Array [],
      "connectionTimeout": 5,
      "dependsOn": Array [],
      "host": "localhost",
      "image": "redis/image:123",
      "networks": undefined,
      "password": "",
      "ports": Object {
        "6379": "6379",
      },
      "props": Object {},
      "responsivenessTimeout": 30,
      "service": "redis",
    },
  },
]
`;
